### DISCUSSION TITLE
iOS libpag Render to CALayer for video overlay

### DISCUSSION THREAD

**Discussion by dilloncc** [Tencent/libpag#2330] [Q&A] [answered]
【Version Info】
libpag (4.3.51)

【Platform Info】
iOS Native

【Expected Behavior】
I am attempting to render a pag animation to a CALayer to display as an overlay in a video. I am able to do this successfully with Lottie when previewing the video with AVPlayer, but exporting the video does not work as expected. I tried the implementation below with libpag, but it never seems to render to the CALayer. 

Attempted Implementation:
```
import UIKit
import AVFoundation
import libpag

public class PAGAnimationLayer: CALayer {
    private var pagView: PAGView!
    private var pagFile: PAGFile!

    public func setupAnimation(name: String, frame: CGRect) {
        self.frame = frame

        // Load the PAG file
        guard let filePath = Bundle.main.path(forResource: name, ofType: "pag") else {
            fatalError("PAG file not found")
        }

        pagFile = PAGFile.load(filePath)

        pagView = PAGView(frame: frame)
        pagView.setComposition(pagFile)
        pagView.setRepeatCount(1) // Set loop mode to play once
        pagView.setSync(true)
        pagView.setVideoEnabled(true)

        self.addSublayer(pagView.layer)
    }

    public func playAnimation() {
        pagView.play()
    }
}
```

Working Lottie Implementation:

```
import UIKit
import AVFoundation
import Lottie

public class LottieAnimationLayer: CALayer {
    private var animationView: LottieAnimationView!

    public func setupAnimation(name: String, frame: CGRect) {
        self.frame = frame
        animationView = LottieAnimationView(name: name)
        animationView.frame = frame
        animationView.contentMode = .scaleAspectFit
        animationView.loopMode = .playOnce
        self.addSublayer(animationView.layer)
    }

    public func playAnimation() {
        animationView.play()
    }
}
```

【Actual Behavior】
The animation is not displaying as an overlay in the video. It doesn't seem to ever render to the CALayer

【Attachments】
[three-arrow.json](https://github.com/user-attachments/files/15595471/three-arrow.json)

**Comment by kevingpqi123** [comment, marked as answer]
In post-editing, live streaming, or shooting scenarios, the rendering content of PAG is combined with the video frames using CVPixelBuffer, which is more efficient than CALayer. Additionally, in these scenarios, PAGSurface, PAGPlayer, and PAGFile are used instead of PAGView. PAGSurface provides an interface to obtain the rendered data of a single frame as CVPixelBuffer, which can be seamlessly integrated with the decoded video content (CVPixelBuffer) to export special effect videos. This usage pattern is the most efficient. However, if you only want to add an effect on top of a video playback component without exporting a special effect video, you can directly add a PAGView. Here, the implementation mechanism of PAG and Lottie differs.

**Reply by dilloncc** [reply]
@kevingpqi123 thank you so much for the quick response. Do you happen to have any hints on how the implementation would look by directly adding a PAGView versus what I have already tried? Thank you again for the help.

**Reply by kevingpqi123** [reply]
The rendering environment in PAGView is derived from CALayer. There may be some issues when directly manipulating the CALayer of PAGView. However, using PAGView as a container eliminates these problems, and the performance difference is insignificant.
